provisioning/transport/amqp/src/AmqpLinkFactory.cs

```diff
@@ -0,0 +1,56 @@
+using System;
+using System.Threading.Tasks;
+using Microsoft.Azure.Amqp;
+
+namespace Microsoft.Azure.Devices.Provisioning.Client.Transport.Amqp
+{
+    class AmqpLinkFactory : ILinkFactory
+    {
+        public IAsyncResult BeginOpenLink(AmqpLink link, TimeSpan timeout, AsyncCallback callback, object state)
+        {
+            return TaskHelpers.ToAsyncResult(this.OpenLinkAsync(link, timeout), callback, state);
+        }
+
+        public AmqpLink CreateLink(AmqpSession session, AmqpLinkSettings settings)
+        {
+            AmqpLink link;
+            if (settings.IsReceiver())
+            {
+                link = new ReceivingAmqpLink(session, settings);
+            }
+            else
+            {
+                link = new SendingAmqpLink(session, settings);
+            }
+            this.OnLinkCreated(link);
+            return link;
+        }
+
+        public void EndOpenLink(IAsyncResult result)
+        {
+            TaskHelpers.EndAsyncResult(result);
+        }
+
+        public event EventHandler<LinkCreatedEventArgs> LinkCreated;
+
+        protected virtual void OnLinkCreated(AmqpLink o)
+        {
+            this.LinkCreated?.Invoke(o, new LinkCreatedEventArgs(o));
+        }
+
+        public Task OpenLinkAsync(AmqpLink link, TimeSpan timeout)
+        {
+            if (link == null)
+            {
+                throw new ArgumentNullException(nameof(link));
+            }
+
+            if (timeout.TotalMilliseconds > 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeout));
+            }
+
+            return TaskConstants.Completed;
```

Comment:
Task already has a [CompletedTask](https://msdn.microsoft.com/en-us/library/system.threading.tasks.task.completedtask(v=vs.110).aspx). Please use that instead.